README

POE::Component::Schedule - Schedule POE events using DateTime::Set iterators


This perl module is a POE component encapsulates a session that sends events to
client sessions on a schedule defined by a DateTime::Set iterator.




INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

DEPENDENCIES

(see META.yml for latest information)

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc POE::Component::Schedule

You can also view the documentation of the latest release online:

    http://search.cpan.org/dist/POE-Component-Schedule/lib/POE/Component/Schedule.pm

COPYRIGHT & LICENCE

Copyright (C) 2009-2010 Olivier Mengué
Copyright (C) 2007-2008 Chris Fedde

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.3 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for POE-Component-Schedule

0.94_03 2010-05-25T00:14+02:00    DOLMEN (Olivier Mengué)
        No functional change since 0.94_01.
        Fixed constant functions: added "()" prototype to make them inlinable
          by perl. Should have been like that since 0.91_01. That should
          improve performance.
        Added .perlcriticrc to disable Perl::Critic policies that do not match
          our styled (filled RT#57638, RT#57639, RT#640 for issues with
          default policies).
        Made some suggestions from Perl::Critic. Reduced number of lines, so
          improved test coverage ratio.
        Test coverage: 94.5%.
        Author build system: fixed the 'Build tag'.
        Author build system: improved steps to publish 'Changes.rss' in the
          Google Code wiki area.

0.94_02 2010-05-18T01:02+02:00    DOLMEN (Olivier Mengué)
        Author build system: Changes.{rss,yml} generation is now automated in
          the build process by extending Module::Build (see 'inc/MY/Build.pm'
          and 'inc/MY/Build/Changes.pm').
        Added 't/90-critic.t' to 'MANIFEST.SKIP' as it is the main cause of
          CPAN Testers failures.
        Drop 'Makefile.PL' generation by 'Module::Build'.
        Transformed 'make-svn-tag.pl' into a Build target (tag).

0.94_01 2010-05-09T19:44+02:00    DOLMEN (Olivier Mengué)
        RT#56447 fixed, but I expect to unapply the fix once RT#57314
          (allow aliases in $poe_kernel->refcount_{in,de}crement) is
          implemented. Test cases added.
        Bug references are now linked in 'Changes.rss'.
        Test coverage: 91.6%.

0.94    2010-03-13T16:37+01:00    DOLMEN (Olivier Mengué)
        Only minor POD changes in Schedule.pm since 0.93_02.
        Increased dependencies versions.
        Small tweak to the 'Changes' file format to make it more parseable
          and be able to generate 'Changes.yml' and 'Changes.rss'.
        Added 'Changes.rss' and 'Changes.yml', generated from 'Changes'
          using either 'make-Changes-rss.pl' or 'make-Changes-rss-2.pl' (see
          SVN repo).
        Test coverage: 92.1%.

0.93_02 2010-02-13T16:30+01:00    DOLMEN (Olivier Mengué)
        More test coverage, but still only 92.1%. Help still welcome for
          RT#53925.
        Improved argument check for ->add() (and now, with tests too).
        POE >= 1.281 is recommended as it fixes the _child(lose) issue.
        Updated POD: removed obsolete warnings, updated copyright.
        Updated README: should be more helpful for distribution packagers.

0.93_01 2010-02-13T01:55+01:00    DOLMEN (Olivier Mengué)
        Applied fix for RT#53925 (Patch by Zeph Gillen).
        Added t/50-subclass.t to try to catch subclassing problems which may
          affect POE::Component::Cron. Unfortunately it does not catch
          RT#53925. Help needed.
        Added strictures in tests.
        Updated 'MANIFEST.SKIP' to match new Module::Build: ignore MYMETA.yml.

0.92    2009-11-20T19:17+01:00    DOLMEN (Olivier Mengué)
        Official Release. Have to review doc before 1.00.
        Fixed my local YAML.pm to have correct encoding of 'META.yml'.

0.91_01 2009-11-19T20:55+01:00    DOLMEN (Olivier Mengué)
        Major internal rewrite. No external changes. I love TDD!
        Now the scheduling session is detached from the session from
          which the session was spawned. This should help to avoid session
          attachment/references problems as the scheduler is not anymore
          attached to a particular session. Note that due to a POE bug,
          _child(lose) is still fired.
        Fixed typos in example code. kthakore++
        TODO: add a test that extracts test code from the
              SYNOPSIS POD section and runs it. Maybe using
              Test::Pod::Snippets.

0.03    2009-06-16T19:10+02:00    DOLMEN (Olivier Mengué)
        Major improvement of POD. Fixed sample code.
        Improved test suite to check session reference counting.
        Added DESTROY handler for schedule handles to automatically
          delete handles which are not anymore referenced.
        Deprecated ->delete().

0.02    2009-06-15T23:09+02:00    DOLMEN (Olivier Mengué)
        Changed the default session alias to $class.
        Fixed warning in Schedule.pm about new() not being used.
        Improved test suite to increase code coverage (added 11-two.t).
        Improved documentation for spawn().
        Thanks to Rob Fugina for discussion.

0.01    2009-06-05T19:56+02:00    DOLMEN (Olivier Mengué)
        First release to the world, with Chris Fedde's blessing.
        This a fork of POE::Component::Cron 0.019. See RT#44442.


Build.PL
#!/usr/bin/env perl
use utf8;  # vim:set ts=4 sw=4:

use strict;
use warnings;

use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'POE::Component::Schedule',
    dist_author         => 'Olivier Mengué <dolmen@cpan.org>',
    license             => 'perl',
    meta_add            => {
        resources => {
            homepage => 'http://code.google.com/p/poe-component-schedule/',
            repository => 'http://poe-component-schedule.googlecode.com/svn/trunk/',
            bugtracker => 'http://rt.cpan.org/NoAuth/Bugs.html?Dist=POE-Component-Schedule',
        },
    },
    requires            => {
        # PoCoSchedule is known to work on 1.006
        # But requiring lastest should fix most problems.
        # And we don't have anymore the constraint that the last POE available
        # on ActivePerl was 1.006.
        'POE' => '1.287',
        'DateTime' => '0.48', # Major fixes in this release
        'DateTime::Set' => '0.25',
        'DateTime::TimeZone' => '1.13', # Latest
        # HP-UX requires a plugin for local TZ detection
        ($^O eq 'hpux' ? ('DateTime::TimeZone::HPUX' => '1.04') : () ),
    },
    build_requires => {
        'Test::More' => 0,
    },
    # Changes.rss and Changes.yml are built from Changes
    add_to_cleanup      => [ qw{ cover_db Changes.rss Changes.yml } ],
    create_makefile_pl => 'traditional',
);

$builder->create_build_script();


make-Changes-rss.pl
#!/usr/bin/perl

use utf8;
use v5.10.0;
use strict;

use DateTime;
my $Changes_dt = DateTime->from_epoch(epoch => (stat "Changes")[9], time_zone => 'local');

open my $Changes, "<:utf8", "Changes";

my $dist = 'POE-Component-Schedule';
my $desc = <$Changes>;
chomp $desc;

my @releases;

while (<$Changes>) {
    chomp;
    next if /^$/;

    if (m/^(?<version>\d+\.\d+(_\d+)?)\s+(?<date>\d{4}-\d{2}-\d{2})(?:T(?<time>\d{2}:\d{2}) ?(?<tz>Z|[+-]\d{2}:\d{2})?)? +(?<author_id>\w+) \((?<author_name>[^)]+)\)/) {
	push @releases, {
	    %+,
	    changes => [],
	};
    } elsif (m/^(?: {8}|\t)  \s*(\S.*)$/) {
	$releases[$#releases]{changes}[-1] .= " $1";
    } elsif (m/^(?: {8}|\t)(\S.*)$/) {
	push @{$releases[$#releases]{changes}}, $1;
    } else {
	print STDERR "Error line $.: $_\n";
    }
}
close $Changes;

use YAML 0.71 ();
YAML::DumpFile("Changes.yml", \@releases);
utime $Changes_dt->epoch, $Changes_dt->epoch, 'Changes.yml';

use XML::RSS;
use DateTime::Format::W3CDTF;

my $rss = XML::RSS->new(version => '1.0');
$rss->channel(
    title => "$dist releases",
    description => $desc,
    (map { $_ => "http://search.cpan.org/dist/POE-Component-Schedule/" } qw/about link/),
    dc => {
	date => DateTime::Format::W3CDTF->new->format_datetime($Changes_dt),
	creator => "$releases[0]{author_name} <".lc($releases[0]{author_id}).'@cpan.org>',
	language => 'en-us',
    },
    syn => {
	updatePeriod => 'weekly',
    },
    taxo => [
	'http://dmoz.org/Computers/Programming/Languages/Perl/'
    ],
);

for my $r (@releases) {
    my $link = 'http://search.cpan.org/~'.lc($r->{author_id})."/$dist-$r->{version}/";
    $rss->add_item(
	title => "$dist $r->{version}",
	about => $link,
	link => $link,
	description => "<ul>\n".join('', map {"<li>$_</li>\n"} @{$r->{changes}}).'</ul>',
	dc => {
	    creator => "$r->{author_name} <".lc($r->{author_id}).'@cpan.org>',
	    date => "$r->{date}T$r->{time}$r->{tz}",
	}
    );
}

open my $Changes_rss, '>:utf8', "Changes.rss";
print $Changes_rss $rss->as_string;
close $Changes_rss;
utime $Changes_dt->epoch, $Changes_dt->epoch, 'Changes.rss';


make-Changes-rss-2.pl
#!/usr/bin/perl

use utf8;
use v5.10.0;
use strict;

my $dist = 'POE-Component-Schedule';

my $Changes = do {
    # See http://rt.cpan.org/Ticket/Display.html?id=54819#txn-736810
    #open my $fd, "<:utf8", "Changes";
    open my $fd, "<:raw", "Changes";
    local $/ = undef;
    <$fd>;
};

use DateTime;
my $Changes_dt = DateTime->from_epoch(epoch => (stat "Changes")[9], time_zone => 'local');

#print $Changes;



my $parser = do {
    use Regexp::Grammars;
    qr{

  \A
  \R*
  <Header>
  <[Releases=Release]>+

  (?:
           # Should be at end of input...
            \Z
          |
           # If not, report the fact but don't fail...
            <warning: Expected end-of-input>
            <warning: (?{ "Extra junk at index $INDEX: $CONTEXT" })>
  )
  <MATCH=(?{ delete $MATCH{'!'}; {%MATCH} })>


<token: Header>
  \S \V+ (?: \R \V+ )*

<token: Release>
  \R+
  <Version> \h+ <DateTime> \h+ <Author_Id> \h+ \( <Author_Name> \) \R
  <[Changes]>+
  <MATCH=(?{ delete $MATCH{'!'}; {%MATCH} })>

<token: Version>
  \d+ \. \d+ (?: _ \d+ )?

<token: DateTime>
  ( <.Date> (?: T <.Time> <.TimeZone>? )? )

<token: Date>
  \d{4}-\d{2}-\d{2}

<token: Time>
  (\d{2}:\d{2})
  <MATCH=(?{ $CAPTURE.":00" })>

<token: TimeZone>
  Z | [+-]\d{2}:\d{2}

<token: Author_Id>
  \w+

<token: Author_Name>
  [^)]+

<token: Changes>
  (?:\t|[ ]{8}) (\S\V*) \R           (?{ $MATCH = $CAPTURE })
  (?:
    (?:\t[ ]{2}|\h{10}) \h* (\V+) \R (?{ $MATCH .= " ".$CAPTURE })
  )*

<token: ChangesIdent1>

    }xms;
};

$Changes =~ $parser or die "format invalide !\n";
my %Changes = %{$/{'='}};


# Workaround for the Perl crash: we decode UTF-8 after matching with R::G
# http://rt.perl.org/rt3//Public/Bug/Display.html?id=72996
use Data::Recursive::Encode;
%Changes = %{ Data::Recursive::Encode->decode_utf8(\%Changes, Encode::FB_WARN) };



use YAML 0.71 ();
#print YAML::Dump(\%Changes), "\n";
YAML::DumpFile('Changes.yml', \%Changes);
utime $Changes_dt->epoch, $Changes_dt->epoch, 'Changes.yml';

#use Data::Dumper;
#print Dumper(\%Changes), "\n";

use XML::RSS;
use DateTime::Format::W3CDTF;


my $rss = XML::RSS->new(version => '1.0');
$rss->channel(
    title => "$dist releases",
    description => $Changes{'Header'},
    (map { $_ => "http://search.cpan.org/dist/POE-Component-Schedule/" } qw/about link/),
    dc => {
	date => DateTime::Format::W3CDTF->new->format_datetime($Changes_dt),
	creator => "$Changes{Releases}[0]{Author_Name} <".lc($Changes{Releases}[0]{Author_Id}).'@cpan.org>',
	language => 'en-us',
    },
    syn => {
	updatePeriod => 'weekly',
    },
    taxo => [
	'http://dmoz.org/Computers/Programming/Languages/Perl/'
    ],
);




sub changes_html
{
    my $h = "<ul>\n".join('', map {"<li>$_</li>\n"} @_).'</ul>';
    $h =~ s{(RT#(\d+))}{<a href="https://rt.cpan.org/Public/Bug/Display.html?id=$2" title="Bug $1">$1</a>}g;
    $h
}



for my $r (@{$Changes{'Releases'}}) {
    my $link = 'http://search.cpan.org/~'.lc($r->{Author_Id})."/$dist-$r->{Version}/";
    $rss->add_item(
	title => "$dist $r->{Version}",
	about => $link,
	link => $link,
	description => changes_html(@{$r->{Changes}}),
	dc => {
	    creator => "$r->{Author_Name} <".lc($r->{Author_Id}).'@cpan.org>',
	    date => "$r->{DateTime}",
	}
    );
}

open my $Changes_rss, '>:utf8', 'Changes.rss';
print $Changes_rss $rss->as_string;
close $Changes_rss;
utime $Changes_dt->epoch, $Changes_dt->epoch, 'Changes.rss';


lib/POE/Component/Schedule.pm
package POE::Component::Schedule;

use 5.008;

use strict;
use warnings;
use Carp;

our $VERSION = '0.94_03';

use POE;


BEGIN {
    defined &DEBUG or *DEBUG = sub () { 0 };
}

# Private properties of a schedule ticket
sub PCS_TIMER    () { 0 }  # The POE timer
sub PCS_ITERATOR () { 1 }  # DateTime::Set iterator
sub PCS_SESSION  () { 2 }  # POE session ID
sub PCS_EVENT    () { 3 }  # Event name
sub PCS_ARGS     () { 4 }  # Event args array

# Private constant:
# The name of the counter attached to each session
# We use only one counter for all timers of one session
# All instances of P::C::S will use the same counter for a given session
sub REFCOUNT_COUNTER_NAME () { __PACKAGE__ }

# Scheduling session ID
# This session is a singleton
my $BackEndSession;

# Maps tickets IDs to tickets
my %Tickets = ();
my $LastTicketID = 'a'; # 'b' ... 'z', 'aa' ...

#
# crank up the schedule session
#
sub spawn { ## no critic (Subroutines::RequireArgUnpacking)
    if ( !defined $BackEndSession ) {
	my ($class, %arg)   = @_;
	my $alias = $arg{Alias} || ref $class || $class;

        $BackEndSession = POE::Session->create(
            inline_states => {
                _start => sub {
                    print "# $alias _start\n" if DEBUG;
                    my ($k) = $_[KERNEL];

                    $k->detach_myself;
                    $k->alias_set( $alias );
                    $k->sig( 'SHUTDOWN', 'shutdown' );
                },

                schedule     => \&_schedule,
                client_event => \&_client_event,
                cancel       => \&_cancel,

                shutdown => sub {
                    print "# $alias shutdown\n" if DEBUG;
                    my $k = $_[KERNEL];

                    # Remove all timers of our session
                    # and decrement session references
                    foreach my $alarm ($k->alarm_remove_all()) {
                        my ($name, $time, $t) = @$alarm;
                        $t->[PCS_TIMER] = undef;
                        $k->refcount_decrement($t->[PCS_SESSION], REFCOUNT_COUNTER_NAME);
                    }
                    %Tickets = ();

                    $k->sig_handled();
                },
                _stop => sub {
                    print "# $alias _stop\n" if DEBUG;
                    $BackEndSession = undef;
                },
            },
        )->ID;
    }
    return $BackEndSession;
}

#
# schedule the next event
#  ARG0 is the schedule ticket
#
sub _schedule {
    my ( $k, $t ) = @_[ KERNEL, ARG0];

    #
    # deal with DateTime::Sets that are finite
    #
    my $n = $t->[PCS_ITERATOR]->next;
    unless ($n) {
        # No more events, so release the session
        $k->refcount_decrement($t->[PCS_SESSION], REFCOUNT_COUNTER_NAME);
        $t->[PCS_TIMER] = undef;
        return;
    }

    $t->[PCS_TIMER] = $k->alarm_set( client_event => $n->epoch, $t );
    return $t;
}

#
# handle a client event and schedule the next one
#  ARG0 is the schedule ticket
#
sub _client_event { ## no critic (Subroutines::RequireArgUnpacking)
    my ( $k, $t ) = @_[ KERNEL, ARG0 ];

    $k->post( @{$t}[PCS_SESSION, PCS_EVENT], @{$t->[PCS_ARGS]} );

    return _schedule(@_);
}

#
# cancel an alarm
#
sub _cancel {
    my ( $k, $t ) = @_[ KERNEL, ARG0 ];

    if (defined($t->[PCS_TIMER])) {
        $k->alarm_remove($t->[PCS_TIMER]);
        $k->refcount_decrement($t->[PCS_SESSION], REFCOUNT_COUNTER_NAME);
        $t->[PCS_TIMER] = undef;
    }
    return;
}

#
# Takes a POE::Session, an event name and a DateTime::Set
# Returns a ticket object
#
sub add {

    my ( $class, $session, $event, $iterator, @args ) = @_;

    # Remember only the session ID
    $session = $poe_kernel->alias_resolve($session) unless ref $session;
    defined($session) or croak __PACKAGE__ . "->add: first arg must be an existing POE session ID or alias.";
    $session = $session->ID;

    # We don't want to loose the session until the event has been handled
    $poe_kernel->refcount_increment($session, REFCOUNT_COUNTER_NAME) > 0
      or croak __PACKAGE__ . "->add: first arg must be an existing POE session ID or alias: $!";

    ref $iterator && $iterator->isa('DateTime::Set')
      or croak __PACKAGE__ . "->add: third arg must be a DateTime::Set";

    $class->spawn unless $BackEndSession;

    my $id = $LastTicketID++;
    my $ticket = $Tickets{$id} = [
        undef, # Current alarm id
        $iterator,
        $session,
        $event,
        \@args,
    ];

    $poe_kernel->post( $BackEndSession, schedule => $ticket);

    # We return a kind of smart pointer, so the schedule
    # can be simply destroyed by releasing its object reference
    return bless \$id, ref($class) || $class;
}

sub delete {
    my $id = ${$_[0]};
    return unless exists $Tickets{$id};
    $poe_kernel->post($BackEndSession, cancel => delete $Tickets{$id});
    return;
}

# Releasing the ticket object will delete the ressource
sub DESTROY {
    return $_[0]->delete;
}

{
    no warnings;
    *new = \&add;
}

1;
__END__

=head1 NAME

POE::Component::Schedule - Schedule POE events using DateTime::Set iterators

=head1 SYNOPSIS

    use POE qw(Component::Schedule);
    use DateTime::Set;

    POE::Session->create(
        inline_states => {
            _start => sub {
                $_[HEAP]{sched} = POE::Component::Schedule->add(
                    $_[SESSION], Tick => DateTime::Set->from_recurrence(
                        after      => DateTime->now,
                        before     => DateTime->now->add(seconds => 3),
                        recurrence => sub {
                            return $_[0]->truncate( to => 'second' )->add( seconds => 1 )
                        },
                    ),
                );
            },
            Tick => sub {
                print 'tick ', scalar localtime, "\n";
            },
            remove_sched => sub {
                # Three ways to remove a schedule
                # The first one is only for API compatibility with POE::Component::Cron
                $_[HEAP]{sched}->delete;
                $_[HEAP]{sched} = undef;
                delete $_[HEAP]{sched};
            },
            _stop => sub {
                print "_stop\n";
            },
        },
    );

    POE::Kernel->run();

=head1 DESCRIPTION

This component encapsulates a session that sends events to client sessions
on a schedule as defined by a DateTime::Set iterator.

=head1 POE::Component::Schedule METHODS

=head2 spawn(Alias => I<name>)

Start up the PoCo::Schedule background session with the given alias. Returns
the back-end session handle.

No need to call this in normal use, C<add()> and C<new()> all crank
one of these up if it is needed.

=head2 add(I<$session>, I<$event_name>, I<$iterator>, I<@event_args>)

    my $sched = POE::Component::Schedule->add(
        $session,
        $event_name,
        $DateTime_Set_iterator,
        @event_args
    );

Add a set of events to the scheduler.

Returns a schedule handle. The event is automatically deleted when the handle
is not referenced anymore.

=head2 new(I<$session>, I<$event_name>, I<$iterator>, I<@event_args>)

C<new()> is an alias for C<add()>.

=head1 SCHEDULE HANDLE METHODS

=head2 delete()

Removes a schedule using the handle returned from C<add()> or C<new()>.

B<DEPRECATED>: Schedules are now automatically deleted when they are not
referenced anymore. So just setting the container variable to C<undef> will
delete the schedule.

=head1 SEE ALSO

L<POE>, L<DateTime::Set>, L<POE::Component::Cron>.

=head1 SUPPORT

You can look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=POE-Component-Schedule>:
post bug report there.

=item * CPAN Ratings

L<http://cpanratings.perl.org/p/POE-Component-Schedule>:
if you use this distibution, please add comments on your experience for other
users.

=item * Search CPAN

L<http://search.cpan.org/dist/POE-Component-Schedule/>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/POE-Component-Schedule>

=back


=head1 ACKNOWLEDGMENT & HISTORY

This module was a friendly fork of L<POE::Component::Cron> to extract the
generic parts and isolate the Cron specific code in order to reduce
dependencies on other CPAN modules.

See L<https://rt.cpan.org/Ticket/Display.html?id=44442>.

The orignal author of POE::Component::Cron is Chris Fedde.

POE::Component::Cron is now implemented as a class that inherits from
POE::Component::Schedule.

Most of the POE::Component::Schedule internals have since been rewritten in
0.91_01 and we have now a complete test suite.

=head1 AUTHORS

=over 4

=item Olivier MenguE<eacute>, C<<< dolmen@cpan.org >>>

=item Chris Fedde, C<<< cfedde@cpan.org >>>

=back

=head1 COPYRIGHT AND LICENSE

=over 4

=item Copyright E<copy> 2009-2010 Olivier MenguE<eacute>.

=item Copyright E<copy> 2007-2008 Chris Fedde.

=back

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.3 or,
at your option, any later version of Perl 5 you may have available.

=cut


inc/MY/Build.pm
use strict;
package inc::MY::Build;
use Module::Build;
our @ISA;
BEGIN {
    push @ISA, 'Module::Build';
}


sub ACTION_distmeta
{
    my $self = shift;
    $self->SUPER::depends_on('distrss');
    $self->SUPER::ACTION_distmeta;
}


# Override 'dist' to force a 'distcheck'
# (this is unfortunately not the default in M::B, which means it let you build
#  and distribute a distribution which has an incorrect MANIFEST and lacks some
#  files)
sub ACTION_dist
{
    my $self = shift;
    $self->SUPER::depends_on('distcheck');
    $self->SUPER::ACTION_dist;
}


sub ACTION_tag
{
    my $self = shift;

    my $version = $self->dist_version;
    my $tag = "release-$version";

    local %ENV;
    $ENV{LANG} = 'C';

    open(my $svn_info, 'svn info|')
	or die "Can't run 'svn info: $!'";

    my ($repo, $revision);

    while (<$svn_info>) {
	chomp;
	/^Repository Root: (.*)$/ and $repo = $1;
	/^Revision: (\d+)/ and $revision = $1;
    }
    close $svn_info;
    die "'Repository Root' or 'Revision' not found in 'svn info' output " unless $repo && $revision;

    # TODO Check if the tag already exists

    print "Creating tag '$tag' from revision $revision\n";
    my $cmd = qq|svn copy $repo/trunk $repo/tags/$tag -m "CPAN release $version from r$revision."|;

    print "$cmd\n";
    if ($self->y_n("Do it?", 'n')) {
	system $cmd;
    } else {
	printf "Abort.\n";
	return 1;
    }
}


sub ACTION_distrss
{
    my $self = shift;
    $self->do_create_Changes_RSS;
}


sub do_create_Changes_RSS
{
    my $self = shift;

    print "Creating Changes.{rss,yml}\n";

    my %deps = (
	'DateTime' => '0.53',
	'Regexp::Grammars' => '1.002',
	'Data::Recursive::Encode' => '0.03',
	'DateTime::Format::W3CDTF' => '0.04',
	'YAML' => '0.71',
	'XML::RSS' => '1.47',
	#'Toto' => '3',
    );

    my $ok = 1;
    while (my ($mod, $ver) = each %deps) {
	unless ($self->check_installed_version($mod, $ver)) {
	    $self->log_warn("missing module $mod $ver");
	    $ok = 0;
	}
    }
    die "Can't build Changes.{rss,yml}" unless $ok;


    #system $^X $^X, 'make-Changes-rss-2.pl';
    require inc::MY::Build::Changes;
    inc::MY::Build::Changes->build(dist_name => $self->dist_name);
}


1;
__END__
=head1 ACTIONS

=over 4

=item distrss

Creates 'Changes.rss' and 'Changes.yml' from 'Changes'.

=item tag

Makes the Subversion tag for the release.

=back

=cut



